### DISCUSSION TITLE
Is it possible to run a matrix strategy job async

### DISCUSSION THREAD

**Discussion by GavinCameron** [community/community#26809] [Actions]
<p>I have a template workflow yml file with 2 jobs, each with the same matrix strategy. The second job depends on the first with <code>needs</code>. I’d like to know if there’s a way to run each value in the matrix async so that the second job doesn’t require all matrix values to complete in the first job before moving on to the second one?</p>
<p>Example main.yml:</p>
<pre><code class="lang-auto">on: [push]

jobs:
  ssql:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        FRUIT: ["apple", "pear"]
    name: Get data
    steps:
      - name: Checkout cum-rev repo
        uses: actions/checkout@v2 # Defaults to current repo - check out current repo
      - name: Checkout ds-ssql-gh-action
        uses: actions/checkout@v2
        with:
          repository: ourorg/ds-ssql-gh-action
          token: ${{ secrets.WORKFLOW_TOKEN }}
          ref: main
          path: './ds-ssql-gh-action'
      - name: Run query and save data
        uses: ./ds-ssql-gh-action
      - name: Upload data as artifact
        uses: actions/upload-artifact@v2
        with:
          name: ${{ matrix.FRUIT }}-data
          path: output_data/data.csv

  Rtrain:
    runs-on: ubuntu-latest
    name: Train model
    needs: ssql
    strategy:
      matrix:
        FRUIT: ["apple", "pear"]
    steps:
      - name: checkout current repo
        uses: actions/checkout@v2 # Defaults to current repo - check out current repo
      - name: Retreive data from ssql job 
        uses: actions/download-artifact@v2
        with:
          name: ${{ matrix.FRUIT }}-data
      - name: Just check csv file is there
        run: ls -l
      - name: Fit model and save as RDS
        uses: ./Rtrain # uses an action in this directory
      - name: Check the model is there
        run: ls -l
      - name: Upload model as artifact
        uses: actions/upload-artifact@v2
        with:
          name: ${{ matrix.FRUIT }}-Model
          path: ${{ matrix.FRUIT }}-Model.RDS
</code></pre>
<p>Where I use <code>needs</code> in the second job, I really mean ‘needs the ssql job for the corresponding fruit, not all jobs in ssql to complete first’.</p>
<p>Suppose the ssql job for ‘apple’ takes 3 times as long as for ‘pear’. I would like the next job, Rtrain, to proceed once the ssql job for ‘pear’ is done without waiting for the ‘apple’ one to complete.</p>
<p>Is this possible?</p>

**Comment by laughedelic** [comment]
<p>For the <code>needs</code> list you need to provide a job ID. You could try to get it for each job of the matrix by adding this step to the first job (<code>ssql</code>):</p>
<pre><code class="lang-yaml">      - name: Print current job ID
        run: echo '${{ github.job }}'
</code></pre>
<p>(see docs on the <a href="https://docs.github.com/en/actions/learn-github-actions/contexts#github-context" rel="noopener nofollow ugc"><code>github</code> context</a>)</p>
<p>Then try to use ID of the concrete job you are interested in in the <code>needs</code> list of the subsequent job.</p>
<p>I don’t know if it will work, but I think it’s worth a try.</p>

**Comment by GavinCameron** [comment]
<p>Presumably here I would need to pass the <code>{{ github.job }}</code> from the first to the second job, perhaps using an artifact. Was that the flow you had in mind here?</p>

**Comment by laughedelic** [comment, marked as answer]
<p>No, you don’t need any artifacts. <code>${{ github.job }}</code> is just a string. You add that step to print it, run your workflow, see the value in the logs, copy it, and then you can remove that step and use the found value in the next job’s <code>needs</code> list.</p>

**Reply by mregen** [reply]
Hi @laughedelic , the `${{ github.job }}` value is the same for me for every job of the matrix if I print it. Did this work for @GavinCameron ?

**Reply by rajatjindal** [reply]
hi  @mregen  did you find a solution for this. I am also in the need of a similar workaround.

**Reply by mregen** [reply]
unfortunately I found no way to wait for a matrix job, only for all jobs when finished.

**Comment by GavinCameron** [comment]
<p>Ah I see, so it’s expected the job id will always be the same, even on subsequent daily runs. OK, gonna try this, thank you!</p>
